### PR TITLE
Feat(blobs/sharedLink) - Add Shared Link Tab to Blobs

### PR DESCRIPTION
**Pull request by sedouard**
- SAS generation under 'SAS' tab
- Modal Tab view -> Modal Tab component
- Store.get -> Store.attr for deprecation
- Blob Component -> Blob Partial + Blob Controller
- Blob Component Tests -> Blob Controller Tests
- SAS Shared Link Tests

### CODE REVIEW

File: app/controllers/explorer.js [hunk 1]
```diff
@@ -10,9 +10,10 @@ import stringResources from '../utils/string-resources';
 export default Ember.Controller.extend({
     // Services & Aliases
     // ------------------------------------------------------------------------------
-    needs: ['notifications'],           // Todo: This looks like a bug - it's neccessary, but shouldn't be
+    needs: ['notifications', 'blob'],           // Todo: This looks like a bug - it's neccessary, but shouldn't be
```
Comment (felixrieseberg) [inline app/controllers/explorer.js]: This shouldn't be necessary (the injection is enough, `needs` is deprecated)

File: app/controllers/blob.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+    actions: {
+        /**
+         * Save properties for Blob model
+         */
+        setProperties: function () {
+            if (this.get('isLocked')) {
+                return;
+            }
+
```
Comment (felixrieseberg) [inline app/controllers/blob.js]: Bonus newline :)

File: app/controllers/blob.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            this.set('showProperties', false);
+            Ember.run.scheduleOnce('afterRender', this, () => {
+                Ember.$('select').material_select();
+            });
+        },
+
+        generateSAS: function () {
+            this.get('model').getLink(this.get('sasExpiration'),
+                this.get('selectedSASOption'))
+            .then(SAS => {
```
Comment (felixrieseberg) [inline app/controllers/blob.js]: `.then(SAS => this.set('SAS', SAS));

**Comment by felixrieseberg** [comment]
This is awesome, but given that `ObjectControllers` are deprecated (which is what we're doing with the blob controller here), isn't there a way to do this with a component? It's kinda sad that we have to rollback something that is future-proof to something deprecated..

**Comment by felixrieseberg** [comment]
Looking at it, I don't think there's any need to go back to a controller. Would you mind if I took this and just implemented it into the component?

File: app/controllers/explorer.js (hunk 1 above)
Comment (sedouard) [inline app/controllers/explorer.js]: i know, but it is necessary for some reason as your comment suggests

**Comment by sedouard** [comment]
`This is awesome, but given that ObjectControllers are deprecated (which is what we're doing with the blob controller here)`

We don't use object controllers. According to [the doc](http://emberjs.com/deprecations/v1.x/#toc_objectcontroller), we're using the right thing.

I basically un-did the change of making this a component, because I think that there will be more functionality than is appropriate in a component. The docs say components are for reusable, simple single action things which translates clicks/actions to meaningful controller/template actions. However having a single component for a modal view which will get a bit more complicated ( User-defined blob metadata will be added at some point) doesn't seem appropriate.

I primarily like this way because of consistent testing patterns on the blob controller as we've done with other controllers for functional, headless testing - and we don't lose any previously written tests.

I'm following the same pattern as the `notifications` controller which I think is working out pretty well. We don't hit any MORE deprecations, just  the ones hanging around still (some with serializer, some with the `select` view)

**Comment by felixrieseberg** [comment]
Actually, the notifications should be a service now (I know, you suggested that all along). We don't call the blob controller an ObjectController, but it still is one. In fact, that's how I was made - as an ItemController for the Array.

I'm happy to do all the work here, if that helps. But controllers for "things" / "items" are no longer a thing.

**Comment by sedouard** [comment]
`We don't call the blob controller an ObjectController, but it still is one.`

I can do the work required - better that I learn it myself. But what I don't understand is how we can declare a controller as a subclass of Ember.Controller, yet it become an Ember.ObjectController. Is it because of the this little nugget in the deprecation doc?

"Opposite to 1.11's ObjectController deprecation, if a controller is not explicitly defined, but instead is being auto-generated by the framework, it will not throw a deprecation message even if the proxying behavior is being used."

Which still doesn't make sense to me because the controller is being explicitly defined....

**Comment by felixrieseberg** [comment]
Thumb rule: if the controller doesn't have its own route, it's baaaaaad. I hate it too, but I'm basically treating this whole project as my "ugh I have to learn it at some point" project. I guess the idea is that everything that doesn't exactly fit a strict MVC goes into a service or a (web) component.

**Comment by sedouard** [comment]
Oh well I just did a bit of research and apparently all use of ember controllers will eventually go away (although [this video](https://www.youtube.com/watch?v=U1HEzVWseng) says they should have been deprecated in v1.13 turns out that hasn't been done just yet I guess).

Alright then I guess components & services it is.

**Comment by sedouard** [comment]
FYI - I'm gonna bring the `notifications` controller out into a service -- so we don't do 2x work

**Comment by felixrieseberg** [comment]
You'd be a hero :+1:

**Pull request merged**